### ISSUE TITLE
Drag a wide vector 

### ISSUE THREAD

**Issue by akovalov** [mousebird-consulting-inc/WhirlyGlobe#1514]
What would be the best way to make a vector line draggable by a long press gesture?

I have a route line add by `addWideVectors` and when drag by gesture remove added `MaplyComponentObject` and add a new changed route line by `addWideVectors`. But as far as I do it on every gesture move it loads cpu a lot and thus is laggy. Also looks like sometimes vector object not gets removed and stays on the map, but not sure yet.

Is there anything like this in AutoTester or could you please head me in the right direction? Thank you.

**Comment by mousebird** [comment]
When adding and removing objects you can specify the mode at the end.  You want Current, so it runs on the current thread.

**Comment by akovalov** [comment]
@mousebird yes, with current thread mode works much better. Also I used a background thread on my own for it, so moved code to main thread as well. Thank you.

**Comment by akovalov** [comment]
@mousebird I have a performance issue with a very long wide vector. The vector is updated as following:

```
let maplyCoordinates = coordinates.map { $0.maplyCoordinate }
let vecObj = MaplyVectorObject(lineString: maplyCoordinates, numCoords: Int32(maplyCoordinates.count), attributes: nil)

switch type {
case .map:
    vecObj.subdivide(toFlatGreatCircle: 0.001)
case .globe:
    vecObj.subdivide(toGlobeGreatCircle: 0.001)
}

if let compObj = self.mapComponentObjs.routeAhead {
    self.maplyVC.remove([compObj], mode: .current)
}
self.mapComponentObjs.routeAhead = self.maplyVC.addWideVectors([vecObj], desc: [
    kMaplyColor: R.color.map.routeFill()!,
    kMaplyVecWidth: (self.zoomLevel >= .two ? 9 : 6) * 2,
    kMaplyDrawPriority: WhirlyDrawPriority.routeAhead,
    kMaplyWideVecJoinType: kMaplyWideVecMiterJoin
], mode: .current)
```

Where `maplyCoordinates` could have 62,000 coordinates and above. 
When this code is performed synchronously on the main thread on the gesture moving the allocated memory steadily increase up to app crash due to no enough memory. This does not help when draw such long line just once for example or with time intervals. Also, when for example RAM memory used increased to 1.5GB and then remove this vector line from the map completely - the memory was not reduced, it still stayed as 1.5GB. So looks like some memory is not freed up if redraw vector multiple times in a row (sequentially, not in parallel). 

Is there a function to free allocated memory to call manually? 
Or do you have any ideas?
Thank you!

**Comment by TimSylvester-mousebird** [comment]
I do see something similar with very large vector objects.  By elimination, it appears to be related to the `subdivide` call.  Oddly, though, I haven't been able to get any callstacks for allocations on the iPad, and the leak doesn't seem to occur on the simulator.

**Comment by TimSylvester-mousebird** [comment]
On further inspection, I think the leak reports I saw were false-positives.  If you can provide a case that triggers the problem, I'll take another look.

<img width="1002" alt="image" src="https://user-images.githubusercontent.com/71895881/158909449-508c93c5-b957-4cc4-85ee-68237151ecb0.png">

**Comment by akovalov** [comment]
@TimSylvester-mousebird here is some data from Allocations Xcode Instruments. Let me know please if I can provide something more specific here.

<img width="718" alt="Screenshot 2022-03-18 at 12 55 22" src="https://user-images.githubusercontent.com/9693643/158991061-acc8a87a-a3d8-4a0a-879c-9a27172523e8.png">

"VM: IOAccelerator" memory is constantly growing, looks like issue is located here.

<img width="1391" alt="Screenshot 2022-03-18 at 12 56 37" src="https://user-images.githubusercontent.com/9693643/158991122-3bd93485-25db-4b85-99b0-6a0ae16ad9bf.png">

<img width="1395" alt="Screenshot 2022-03-18 at 12 57 27" src="https://user-images.githubusercontent.com/9693643/158991160-e04a340b-9a5c-464a-bea9-e6b25d57ebd4.png">

**Comment by TimSylvester-mousebird** [comment]
I've never been entirely clear what that is, but I think it has to do with the CPU/graphics memory interface.

It doesn't seem likely that it's something as simple as we're leaking Metal resources, but perhaps with larger-than-normal allocations you're getting fragmentation in the Metal heaps?

**Comment by akovalov** [comment]
@TimSylvester-mousebird thank you for your suggestions and time spent to look into this issue. 
I've ended up with removing subdivide calls and optimizing own code on updating the line on long gesture moving. 
This memory issue happened only when the route line was extra huge, so I've changed how to work with such long line to not redraw it on gesture moving. 
And also I think it was related to how I work with threads to update the line. 
So now I have it working good enough and closing this ticket. Thank you!

**Comment by akovalov** [comment]
@TimSylvester-mousebird one question though, what did you mean by "Change to performance wide vectors to eliminate the CPU-based line-to-polygon code"? I've used `addWideVectors` .

**Comment by TimSylvester-mousebird** [comment]
I'm referring to using `kMaplyWideVecImpl: kMaplyWideVecImplPerf` in the descriptor dictionary.  See the wide vectors test case in the AutoTester app for examples.

This selects a version where most of the computation (finding the triangles which cover the given line) is done on the GPU, reducing CPU load and the amount of data being sent to the GPU.  Some of the features like offsets, texturing, and bevel joins are not yet supported, however.  I happen to be in the middle of adding some of those features so, if you do need them, they should be available soon.

**Comment by akovalov** [comment]
@TimSylvester-mousebird tried it and the line visually looks differently than without `kMaplyWideVecImplPerf` and it does not change situation with memory usage, so I'll stay with how it was. Ok, thanks a lot!

**Comment by TimSylvester** [comment]
Significant updates to wide vectors were merged on Friday (#1534) so consider taking another look when you next update.

Another thing I thought of regarding memory is select-ability; if the objects have `kMaplySelectable: true`, the vector object is retained for hit-testing.  But if you don't need selection, or you can handle the hit-testing yourself on geometry that you're already retaining outside the toolkit, that isn't necessary.
